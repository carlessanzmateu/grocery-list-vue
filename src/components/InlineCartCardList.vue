<template>
  <section class="inline-cart-card-list grid grid-cols-1 p-4">
    <h3 class="pb-3">Cart</h3>
    <inline-cart-card
      class="col-span-1"
      v-for="(item, index) in items"
      :key="index"
      :item="item"
      @add-item="addItem"
      @remove-item="removeItem"
    />
    <checkout
      :total-amount="totalAmount"
    />
  </section>
</template>

<script>
import Checkout from '@/components/Checkout.vue'
import InlineCartCard from '@/components/InlineCartCard'

export default {
  name: 'InlineCartCardList',
  props: {
    items: {
      type: Array,
      required: true
    }
  },
  components: {
    InlineCartCard,
    Checkout
  },
  computed: {
    totalAmount () {
      if (!this.items.length) {
        return 0
      }
      let totalAmount = 0

      this.items.forEach(item => {
        totalAmount += item.getItemTotalPrice()
      })

      return totalAmount
    }
  },
  methods: {
    addItem (item) {
      this.$emit('add-item', item)
    },
    removeItem (item) {
      this.$emit('remove-item', item)
    }
  }
}
</script>
