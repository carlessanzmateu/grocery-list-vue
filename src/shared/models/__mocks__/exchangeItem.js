const mockExchangeItem = {
  id: '1',
  // eslint-disable-next-line
  image_url: 'mock.png',
  stock: 1,
  productName: 'some name',
  price: 10,
  productDescription: 'Some description',
  favorite: 0
}
