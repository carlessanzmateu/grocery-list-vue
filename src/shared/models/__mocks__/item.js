class MockItem {
  constructor() {
    this.$id = '1'
    this.$image = 'mock.png'
    this.$stockReadyToBuy = 0
    this.$stock = 1
    this.$name = 'some name'
    this.$price = 10
    this.$description = 'Some description'
    this.$favorite = false
  }

  getId () {
    return this.$id
  }

  getImage () {
    return this.$image
  }

  initSelectedStock () {
    this.$stockReadyToBuy = 0
  }

  getSelectedStock () {
    return this.$stockReadyToBuy
  }

  getStock () {
    return this.$stock
  }

  getStockLeft () {
    return this.getStock() - this.getSelectedStock()
  }

  canIncreaseSelectedStock () {
    return this.getSelectedStock() < this.getStock()
  }

  canDecreaseSelectedStock () {
    return this.getSelectedStock() > 0
  }

  increaseSelectedStock () {
    if (!this.canIncreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy++
  }

  decreaseSelectedStock () {
    if (!this.canDecreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy--
  }

  getName () {
    return this.$name
  }

  getPrice () {
    return this.$price
  }

  getItemTotalPrice () {
    return this.getSelectedStock() * this.getPrice()
  }

  getDescription () {
    return this.$description
  }

  buildFavorite (favorite) {
    this.$favorite = (favorite === '1')
  }

  getIsFavorite () {
    return this.$favorite
  }

  setAsFavorite () {
    this.$favorite = true
  }

  removeFromFavorite () {
    this.$favorite = false
  }

  toggleFavorite () {
    this.$favorite = !this.$favorite
  }
}

export { MockItem }
