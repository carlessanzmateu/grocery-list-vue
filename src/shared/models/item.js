export default class Item {
  $id
  $image
  $stockReadyToBuy
  $stock
  $name
  $price
  $description
  $favorite

  constructor (rawItem) {
    this.$id = rawItem.id
    this.$image = rawItem.image_url
    this.$stock = rawItem.stock
    this.$name = rawItem.productName
    this.$price = rawItem.price
    this.$description = rawItem.productDescription
    this.buildFavorite(rawItem.favorite)
    this.initSelectedStock()
  }

  getId () {
    return this.$id
  }

  getImage () {
    return this.$image
  }

  initSelectedStock () {
    this.$stockReadyToBuy = 0
  }

  getSelectedStock () {
    return this.$stockReadyToBuy
  }

  // setSelectedStock (selectedStock) {
  //   this.$stockReadyToBuy = selectedStock
  // }

  getStock () {
    return this.$stock
  }

  setStock (stock) {
    this.$stock = stock
  }

  getStockLeft () {
    return this.getStock() - this.getSelectedStock()
  }

  canIncreaseSelectedStock () {
    return this.getSelectedStock() < this.getStock()
  }

  canDecreaseSelectedStock () {
    return this.getSelectedStock() > 0
  }

  increaseSelectedStock () {
    if (!this.canIncreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy++
  }

  decreaseSelectedStock () {
    if (!this.canDecreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy--
  }

  getName () {
    return this.$name
  }

  getPrice () {
    return this.$price
  }

  getItemTotalPrice () {
    return this.getSelectedStock() * this.getPrice()
  }

  getDescription () {
    return this.$description
  }

  buildFavorite (favorite) {
    this.$favorite = (favorite === '1')
  }

  getIsFavorite () {
    return this.$favorite
  }

  setAsFavorite () {
    this.$favorite = true
  }

  removeFromFavorite () {
    this.$favorite = false
  }

  toggleFavorite () {
    this.$favorite = !this.$favorite
  }
}
