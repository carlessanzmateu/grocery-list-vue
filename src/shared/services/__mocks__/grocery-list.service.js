export default {
  getAllGroceries: jest.fn(),
  getAllFavorites: jest.fn(),
  updateGrocery: jest.fn()
}
