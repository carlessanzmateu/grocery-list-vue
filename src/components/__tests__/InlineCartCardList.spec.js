import { shallowMount } from '@vue/test-utils'
import InlineCartCardList from '@/components/InlineCartCardList.vue'
import { MockItem } from '@/shared/models/__mocks__/item'

describe('InlineCartCardList', () => {
  let wrapper
  function createWrapper() {
    return shallowMount(InlineCartCardList, {
      propsData: {
        items: [new MockItem(), new MockItem()],
      }
    })
  }

  it('should create', () => {
    wrapper = createWrapper()
    expect(wrapper.vm).toBeTruthy()
  })

  it('should match html', () => {
    wrapper = createWrapper()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should have addItem method', () => {
    wrapper = createWrapper()

    expect(wrapper.vm.addItem).toBeDefined()
  })

  it('should emit add-item event when call addItem method', () => {
    wrapper = createWrapper()

    wrapper.vm.addItem()

    expect(wrapper.emitted()['add-item'])
  })

  it('should have removeItem method', () => {
    wrapper = createWrapper()

    expect(wrapper.vm.removeItem).toBeDefined()
  })

  it('should emit remove-item event when call removeItem method', () => {
    wrapper = createWrapper()

    wrapper.vm.removeItem()

    expect(wrapper.emitted()['remove-item'])
  })

  it('should set total amount as 0 when has no items', () => {
    const expectedTotalAmount = 0
    wrapper = createWrapper()

    wrapper.setData({
      items: []
    })

    expect(wrapper.vm.totalAmount).toBe(expectedTotalAmount)
  })
})
