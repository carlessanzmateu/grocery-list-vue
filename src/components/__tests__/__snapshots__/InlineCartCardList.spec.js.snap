// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineCartCardList should match html 1`] = `
<section
  class="inline-cart-card-list grid grid-cols-1 p-4"
>
  <h3
    class="pb-3"
  >
    Cart
  </h3>
   
  <inline-cart-card-stub
    class="col-span-1"
    item="[object Object]"
  />
  <inline-cart-card-stub
    class="col-span-1"
    item="[object Object]"
  />
   
  <checkout-stub
    totalamount="0"
  />
</section>
`;
